[server] Retention emails

mngshm opened the pull request.

mngshm committed:
[server](SendFamilyNudge) method to get list of all subscribed users

mngshm committed:
[server](SendFamilyNudge) Modify repo func to get subscribed users without families

mngshm committed:
[server](SendFamilyNudge) controller method to send emails to user without families after 30 days of sub

mngshm committed:
[server](SendFamilyNudge) record email nudge time in notification_history table

mngshm committed:
[server](StorageExceedingAlerts) refactor/modify existing method to return storage exceeded/exceeding users on the param basis

mngshm committed:
[server](StorageExceedingAlerts) struct to loop through slices of set of users depending on storage consumed

mngshm committed:
[server](StorageExceedingAlerts) handle bonus storage separately

mngshm committed:
[server](ServerExceedingAlerts) fix improper SetlastNotificationTimeToNow calls

mngshm committed:
[server](StorageExceedingAlerts) Better struct name

mngshm committed:
[server](StorageExceedingAlerts) repo func to delete record from notification_history table

mngshm committed:
[server](StorageExceedingAlerts) Delete user's record from notification_history on Subscription Update

mngshm committed:
[server](StorageExceedingAlerts) Minor tweaks in multiple files

mngshm committed:
Update Comment

mngshm committed:
Comment

mngshm committed:
Family nudge email template

mngshm committed:
90% storage exceeded email template

mngshm committed:
update subject, notification ID's

mngshm committed:
turn off OnFirstFIleUpload email trigger

ua741 commented on server/pkg/controller/billing.go:
```diff
@@ -366,6 +370,13 @@ func (c *BillingController) UpdateSubscription(r ente.UpdateSubscriptionRequest)
 	if err != nil {
 		return stacktrace.Propagate(err, "")
 	}
+
+	// Delete users entry from notification_history table on subscription update
+	err = c.NotificationHistory.DeleteLastNotificationTime(r.UserID, "90_percent_consumed")
+	if err != nil {
+		return stacktrace.Propagate(err, "")
+	}
+
```
Is this flow also triggered when a normal user changes their subscription or just when support team updates the subscription?

mngshm replied:
Apparently, there are no checks checking if the update was triggered by the user themselves (support is also an admin account if I'm not wrong), so this change will get triggered in both cases. And ideally, I'm guessing this change in the table has to be triggered in any case when the subscription has changed because anyhow to reset the status of 90% consumption.

ua741 replied:
If we check the usage of `UpdateSubscription`, we can identify that this method is only called when support updates the subscription.

Assuming if we only want to reset the alerts on subscription change, we should look at where all we are calling `ReplaceSubscription` method in the billing repo.

Also, if a users goes above 90%, deletes some data, and then again goes above 90, we don't want to send email again?

mngshm replied:
>Assuming if we only want to reset the alerts on subscription change, we should look at where all we are calling ReplaceSubscription method in the billing repo. 

Ack, i'll check for these points and make necessary changes.

>Also, if a users goes above 90%, deletes some data, and then again goes above 90, we don't want to send email again?

Ideally, this email notification is added to tell the user that they're going to exceed the storage quota very sooner. So, yes, it will send an email if the user exceeds 90%, deletes some data and then again crosses the threshold.

mngshm committed:
remove unnecessary log statement

ua741 commented on server/pkg/repo/user.go:
```diff
@@ -280,17 +310,17 @@ func (repo *UserRepository) GetPublicKey(userID int64) (string, error) {
 		INNER JOIN subscriptions 
-		ON users.user_id = subscriptions.user_id AND usage.storage_consumed > subscriptions.storage AND users.encrypted_email IS NOT NULL AND users.family_admin_id IS NULL;
-	`)
+		ON users.user_id = subscriptions.user_id AND usage.storage_consumed >= (subscriptions.storage * $1 / 100.0) AND users.encrypted_email is not null AND users.family_admin_id is NULL;
```
`* $1 / 100.0` Is this intentional?

mngshm replied:
Some AI bad code which missed through my eyes while re-reviewing code.

ua741 replied:
Understood.
I don't think equal will work as user storage might never reach exact 90%. Even if it reaches, the cron runs at fixed interval, it would increase after that.

mngshm replied:
[0f24ba0](https://github.com/ente-io/ente/pull/5550/commits/0f24ba01f523fbb87cb7bdf2d57edecbf14f76a4)

Refactored

ua741 commented on server/pkg/controller/email/email_notification.go:
```diff
@@ -171,33 +182,113 @@ func (c *EmailNotificationController) SendStorageLimitExceededMails() {
+				break
+			}
+
+			thirtyDays := 30 * 24 * t.Hour
+			creationTime := t.Unix(0, user.CreationTime)
+			timeSinceCreation := t.Since(creationTime)
+
+			if timeSinceCreation >= thirtyDays {
```
We can this filter in `GetSubscribedUsersWithoutFamily`, can pass a param createdBeforeInDays int64 and use that to get only users with account older than 30 days.

mngshm replied:
Hmm. Probably this is a better way to do it then. In one more type of retention-email I have to send an email to list of users who have signed up half hour ago. Maybe i'll try to get that done in this PR itself.

mngshm replied:
I will give this a try @ua741

mngshm replied:
made requested changes in [ce37e9d](https://github.com/ente-io/ente/pull/5550/commits/ce37e9d573248f78b928bc5c4c7e578b1a3758d7)

mngshm committed:
minor fix

mngshm committed:
Delete entry from  table whenever there is a change in subscription

mngshm committed:
delete last notification time from table after successful execution of ReplaceSubscription

mngshm committed:
modify code to pass timeDuration as parameter to get list of users

previously all users without families were fetched and later 30 days check was being done.
With this modification, in the future custom duration can be passed and depending on that one could get the list of users. The compiler decides if more time has passed dependeing on what is the underlying nanoseconds value of time.Time value. Older dates are less in size compared to Newer Dates. Hence, s.created_at >=  is an appropriate calculation

ua741 commented on server/pkg/controller/email/email_notification.go:
```diff
@@ -171,33 +182,113 @@ func (c *EmailNotificationController) SendStorageLimitExceededMails() {
+		return stacktrace.Propagate(subUsersErr, "Failed to get subscribers")
+	}
+
+	batchSize := 100
+	totalSubUsers := len(subscribedUsers)
+
+	for i := range totalSubUsers {
+		end := i + batchSize
+		if end > totalSubUsers {
+			end = totalSubUsers
+		}
```
What's the motivation behind this batching?

mngshm replied:
We had discussed, we would be batching in case for bulk emails 😅 ?

ua741 commented on server/pkg/repo/user.go:
```diff
@@ -280,17 +313,17 @@ func (repo *UserRepository) GetPublicKey(userID int64) (string, error) {
 		INNER JOIN subscriptions 
-		ON users.user_id = subscriptions.user_id AND usage.storage_consumed > subscriptions.storage AND users.encrypted_email IS NOT NULL AND users.family_admin_id IS NULL;
-	`)
+		ON users.user_id = subscriptions.user_id AND usage.storage_consumed = (subscriptions.storage * $1 / 100) AND users.encrypted_email is not null AND users.family_admin_id is NULL;
```
This ($1/100) change is still present.

mngshm replied:
https://github.com/ente-io/ente/pull/5550#discussion_r2052227392

ua741 commented on server/pkg/controller/email/email_notification.go:
```diff
@@ -171,33 +182,114 @@ func (c *EmailNotificationController) SendStorageLimitExceededMails() {
+		}
+		batchUsers := subscribedUsers[i:end]
+		for _, user := range batchUsers {
+			isFamilyAdmin, err := c.UserRepo.GetFamilyAdminID(user.ID)
+			if err != nil {
+				return stacktrace.Propagate(err, "User is a family admin")
+			}
+			if isFamilyAdmin != nil {
+				break
+			}
+
+			creationTime := t.Unix(0, user.CreationTime)
+			timeSinceCreation := t.Since(creationTime)
+			if timeSinceCreation >= thirtyDaysDuration {
```
Both GetFamilyAdminID and creationTime check are redundant? `GetSubscribedUsersWithoutFamily` function has already checks/filters for it.

mngshm replied:
Nope, there aren't. I followed our practice of not doing any checks on repository (DB) level code. Hence, all the checks are being done inside the controller layer.

https://github.com/ente-io/ente/blob/f609cef79e2e040c4c1e9ec929c4b1be1b5a1a29/server/pkg/repo/user.go#L133

ua741 commented on server/pkg/repo/user.go:
```diff
@@ -128,6 +129,38 @@ func (repo *UserRepository) GetAll(sinceTime int64, tillTime int64) ([]ente.User
 }
 
+// GetSubscribedUsers will return the list of all subscribed.
+func (repo *UserRepository) GetSubscribedUsersWithoutFamily(duration string) ([]ente.User, error) {
```
So, the created_at does not get updated when user switches from free to paid plan.
```go
func (repo *UserRepository) GetSubscribedUsersWithoutFamily(accountAgeInDays int) ([]int64, error) {
	rows, err := repo.DB.Query(`SELECT u.user_id, u.encrypted_email, u.email_decryption_nonce, u.email_hash, s.created_at 
	FROM subscriptions s 
	INNER JOIN users u ON s.user_id = u.user_id 
	WHERE u.creation_time <= $1
	AND u.family_admin_id IS NULL
	AND u.encrypted_email IS NOT NULL
	AND s.product_id != 'free'`, time.MicrosecondBeforeDays(accountAgeInDays))
	
	...
```

mngshm replied:
In this query, the condition happens between `u.creation_time` <= $1. This is not what was intended. The user might chose to not subscribe and consume the free 10GB storage first. Hence, I went ahead with using `s.created_at`. Ideally, the mail has to be sent to those who have crossed 30 days after subscribing and still do not have a family onboarded.

mngshm replied:
Fixed the requested changes in [abe0ef0](https://github.com/ente-io/ente/pull/5550/commits/abe0ef0a0354c674e6ddf00e7b4ef02f386e6da2)

ua741 commented on server/pkg/repo/user.go:
```diff
@@ -128,6 +129,38 @@ func (repo *UserRepository) GetAll(sinceTime int64, tillTime int64) ([]ente.User
+	rows, err := repo.DB.Query(`SELECT u.user_id, u.encrypted_email, u.email_decryption_nonce, u.email_hash, s.created_at 
+	FROM subscriptions s 
+	INNER JOIN users u ON s.user_id = u.user_id 
+	WHERE s.created_at >= $1
+	AND u.family_admin_id IS NULL 
```
s.created_at >= $1 should be s.created_at <= $1 (fixed it in the suggested query) and also added check for deleted users.

ua741 commented on server/pkg/controller/billing.go:
```diff
@@ -53,6 +55,7 @@ func NewBillingController(
 	usageRepo *repo.UsageRepository,
 	storageBonusRepo *storagebonus.Repository,
 	commonBillCtrl *commonbilling.Controller,
+	notificationHistory *repo.NotificationHistoryRepository,
```
nit: notificationHistoryRepo or just notificaitonRepo. Please rename it at other places too.

mngshm committed:
remove batching

mngshm committed:
remove Dependency injection and functionality for DeleteLastNotificationTime

mngshm committed:
refactor: dedup code

mngshm committed:
fix query, add filter for deleted users, switch temporarily to u.creation_time

mngshm committed:
turn off the cron for SendFamilyNudge temporarily

ua741 committed:
simplify family nudge email

ua741 committed:
Fix typo & minor refactor

ua741 committed:
Fix query & add doc

ua741 committed:
clean up

ua741 committed:
Fix bug in GetUsersWithExceedingStorages

ua741 committed:
Merge branch 'main' into retention-emails

ua741 committed:
Avoid spinning go routines for individual emails

ua741 committed:
rename
